packaging/release.py

```diff
@@ -0,0 +1,1410 @@
+
+    def main(self) -> None:
+        """Main program entry point."""
+        parser = argparse.ArgumentParser(description=__doc__)
+        subparsers = parser.add_subparsers(metavar="COMMAND", required=True)
+
+        for func in self.commands:
+            func_parser = subparsers.add_parser(self._format_command_name(func), description=func.__doc__, help=func.__doc__)
+            func_parser.set_defaults(func=func)
+
+            exclusive_groups = {}
+            signature = inspect.signature(func)
+
+            for name in signature.parameters:
+                if name not in self.arguments:
+                    raise Exception(f"The '{name}' argument, used by '{func.__name__}', has not been defined.")
+
+                if (arguments := self.arguments.get(name)) is None:
+                    continue  # internal use
+
+                arguments = arguments.copy()
+                exclusive = arguments.pop("exclusive", None)
+
+                if exclusive:
+                    if exclusive not in exclusive_groups:
+                        exclusive_groups[exclusive] = func_parser.add_mutually_exclusive_group()
+
+                    command_parser = exclusive_groups[exclusive]
+                else:
+                    command_parser = func_parser
+
+                if option_name := arguments.pop("name", None):
+                    arguments.update(dest=name)
+                else:
+                    option_name = f"--{name.replace('_', '-')}"
+
+                command_parser.add_argument(option_name, **arguments)
+
+        try:
+            # noinspection PyUnresolvedReferences
+            import argcomplete
+        except ImportError:
+            pass
+        else:
+            argcomplete.autocomplete(parser)
+
+        self.parsed_arguments = parser.parse_args()
+
+        try:
+            self.run(self.parsed_arguments.func)
+        except ApplicationError as ex:
+            display.fatal(ex)
+            sys.exit(1)
+
+    def _run(self, func: C, **kwargs) -> None:
+        """Run the specified command, using any provided internal args."""
+        signature = inspect.signature(func)
+        func_args = {name: getattr(self.parsed_arguments, name) for name in signature.parameters if hasattr(self.parsed_arguments, name)}
+        func_args.update({name: value for name, value in kwargs.items() if name in signature.parameters})
+        printable_args = ", ".join(f"{name}={repr(value)}" for name, value in func_args.items())
+        label = f"{self._format_command_name(func)}({printable_args})"
+        failed = False
+
+        display.show(f"==> {label}", color=Display.BLUE)
+
+        try:
+            func(**func_args)
+        except BaseException:
+            failed = True
+            raise
+        finally:
+            if failed:
+                display.show(f"!!! {label}", color=Display.RED)
+            else:
+                display.show(f"<== {label}", color=Display.BLUE)
```

Comment:
Why didn't you go for
```suggestion
            display.show(f"!!! {label}", color=Display.RED)
            raise

        display.show(f"<== {label}", color=Display.BLUE)
```
?

A flag changed in “except” and re-accessed in “finally” almost after the function's control flow is interrupted is probably not the easiest to follow.

Plus, this variant is less nested.

Reply:
I reworked this a few times, and somehow missed this cleanup when doing so. It's simplified now.